src/libexpr/primops.cc

```diff
@@ -3664,6 +3664,84 @@ static RegisterPrimOp primop_hashString({
     .fun = prim_hashString,
 });
 
+static void prim_convertHash(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    auto type = state.forceStringNoCtx(*args[0], pos, "while evaluating the first argument passed to builtins.convertHash");
+    std::optional<HashType> ht = parseHashType(type);
+    if (!ht)
+        state.debugThrowLastTrace(Error({
+            .msg = hintfmt("unknown hash type '%1%'", type),
+            .errPos = state.positions[pos],
+        }));
+
+    auto base = state.forceStringNoCtx(*args[1], pos, "while evaluating the second argument passed to builtins.convertHash");
+    Base hb;
+    if (base == "sri") {
+        hb = SRI;
+    } else if (base == "base16") {
+        hb = Base16;
+    } else if (base == "base32") {
+        hb = Base32;
+    } else if (base == "base64") {
+        hb = Base64;
+    } else {
+        state.debugThrowLastTrace(Error({
+            .msg = hintfmt("unknown hash format '%1%'", base),
```

Comment:
The hash-conversion implementation is shared with that of the CLI `nix-hash` and `nix hash`, which handles errors and prints error messages for the latter tools.

Let's also delegate the error handling to the implementation, and improve the error message there if need be.

Reply:
Update: Sorry that I confused hash format (base) with hash algorithm (type) in the last reply.

I'll add a C++ function `parseHashBase` so that the implementation can be shared between the `builtins.convertHash` primop and the renamed `nix hash convert` interface mentioned  in #8876.

Reply:
Addressed.